### PR TITLE
Create Backend API for Rescue Data Fetching

### PR DESCRIPTION
**Pull request by ryanmarshallmc**
*This is a big ole update!*

We're officially reaching the point when our database has grown to the limits of our current data fetching implementation. What a good problem to have!

### The Problem

Up to this point, we've chosen to simply load all data from our db (within a certain date range) into the frontend, and then sort and join data (ie. adding stops to each rescue) on the frontend. This was super nice, because it meant we could eliminate the need for any "serve side" code, doing everything in the context of the React app. Cool!

However, we're getting to the point where that is simply too much data to load, and setting a firm date range causes issues around the app loading specific rescues before we know if the date is within our current range. TLDR: it's getting messy.

### One possible solution that we're *not* choosing....

We could fetch the specific data we need for each page directly from the frontend, instead of loading all data in advance.

However, because of the relational nature of our data (rescues depend on stops, depend on orgs, depend on locations, etc.), fetching each of these pieces of data sequentially would be super slow, far worse than our load times today.

### A better solution implemented in this PR!

We create a server side API, with specific endpoints built to returns clean structured data. For example, the new `/rescue/:id` API endpoint returns a rescue complete with stops, organizations, locations, and handler data, all in one request.

Pros: we can now abstract all of that data handling/joining/formatting in an isolated context on the server, where it will run fast and have faster round trip times to the actual Firestore database.

Cons: definitely a few.
 - We need to run the server side Firebase Cloud Functions locally using Emulators in order to develop on new API code.
 - We need to debug across 2 environments instead of one.
 - We need to make a separate endpoint for any specific data shape we need in the future.
 
 
 *TLDR:* I think this is a good set of trade-offs for now, and will set us up to scale much better long term. But there are some adjustments we need to make to how we develop locally to make that happen smoothly.

### CODE REVIEW

**Comment by ryanmarshallmc** [comment]
A quick review of how the file system has changed here:

 - The `functions` directory has been renamed `backend`, to better delineate between frontend and backend/"server" code
 - Inside `backend`, we have new subdirectories `environments`, `functions`, and `helpers`, to mirror the same structure we use in the frontend `src` directory
 - Inside `backend/functions`, each directory represents the code for a specific firebase function. `api` is now the biggest, containing all of the various calls the frontend app can make directly to the server (including analytics, calendar events, and data fetching)

**Comment by ryanmarshallmc** [comment]
## TODOs:

 - [ ] Review security credentials for `backend`
 - [ ] Test running locally with `prod` configuration
 - [ ] Add server side error handling with Sentry or similar tool
 - [ ] Frontend QA/testing with new API code
